src/Network/GRPC/MQTT/Proto.hs

```diff
@@ -0,0 +1,672 @@
+
+import Control.Monad.Except (MonadError, throwError)
+
+import Data.Data (Data, cast, eqT, gmapQ)
+import Data.Traversable (for)
+import Data.Foldable (foldrM)
+import Data.Type.Equality ((:~:) (Refl))
+
+import Data.List.NonEmpty qualified as NonEmpty
+import Data.Map qualified as Map
+
+import Language.Haskell.TH (Name)
+
+import Network.GRPC.HighLevel.Generated
+  ( GRPCMethodType (BiDiStreaming, ClientStreaming, Normal, ServerStreaming),
+  )
+
+import Proto3.Suite.DotProto
+  ( DotProto,
+    DotProtoDefinition,
+    DotProtoIdentifier,
+    DotProtoOption,
+    DotProtoServicePart,
+    DotProtoValue,
+    RPCMethod,
+  )
+import Proto3.Suite.DotProto qualified as DotProto
+import Proto3.Suite.DotProto.Internal (CompileError)
+import Proto3.Suite.DotProto.Internal qualified as DotProto
+
+import Proto3.Wire.Decode (ParseError)
+import Proto3.Wire.Decode qualified as Wire
+
+import Relude
+
+import Text.Show qualified as Show
+
+import Turtle qualified
+
+--------------------------------------------------------------------------------
+
+-- Orphan instance @'Data' 'DotProtoValue'@
+import Proto3.Suite.Orphans ()
+
+-- Proto IO --------------------------------------------------------------------
+
+-- | Read the contents of a *.proto file into a 'DotProto' syntax tree.
+--
+-- Raises 'ProtoIOError' 'IO' exceptions if the 'Turtle.FilePath' does not refer
+-- to a valid *.proto file that can be read.
+--
+-- @since 0.1.0.0
+openProtoFileIO :: Turtle.FilePath -> IO DotProto
+openProtoFileIO filepath = do
+  -- Test that the path @filepath@ exists.
+  unlessM (Turtle.testpath filepath) do 
+    throwProtoDoesNotExistIO filepath
+
+  -- Test that @filepath@ points to a file, not a directory.
+  unlessM (Turtle.testfile filepath) do 
+    throwProtoPathIsDirectoryIO filepath
+
+  -- Test that path @filepath@ has a *.proto extension.
+  let isproto = Turtle.hasExtension filepath "proto"
+  unless isproto (throwProtoInvalidExtensionIO filepath)
+
+  -- Read the contents @filepath@, handle any protobuf compile errors should any
+  -- be raised by @importProtoFile@.
+  let file = Turtle.filename filepath
+  let dire = Turtle.directory filepath
+  runExceptT (DotProto.importProto [dire] file file) >>= \case
+    Left err -> throwCompileErrorIO filepath err
+    Right dp -> pure dp
+
+-- Proto IO Errors -------------------------------------------------------------
+
+-- | 'ProtoIOError' represents an IO error that occurs when reading or parsing
+-- the contents of a *.proto file.
+--
+-- @since 0.1.0.0
+data ProtoIOError = ProtoIOError
+  { ioeFilePath :: Turtle.FilePath
+  , ioeErrorTag :: ProtoIOErrorTag
+  }
+  deriving stock (Eq, Ord)
+
+-- | @since 0.1.0.0
+instance Show ProtoIOError where
+  show (ProtoIOError filepath tag) =
+    let locS = "(" ++ Turtle.encodeString filepath ++ "): "
+        tagS = Show.shows tag "."
+     in "protobuf IO error" ++ locS ++ tagS
+  {-# INLINE show #-}
```

Comment:
I generally expect `show` to give me the standard derived string output, unless it is trying to support some kind of abstraction barrier.

For custom `Exception` types I think it's better to derive `Show` and implement this pretty printing in `displayException`.